### PR TITLE
[FEAT] Restricted Pancakeswap V2 Strategies

### PR DESCRIPTION
**Pull request by abraham-alpaca**
## Description
This is a PR for creating restricted PancakeSwap V2 strategies. The difference between typical strategies and the restricted one is that the restricted one will whitelist some workers for allowing only those workers to call an `execute` function. all base and farming tokens will be derived from those workers as well. Hence, an issue of an incorrect injected token will be solved.

### CODE REVIEW

File: contracts/6/protocol/strategies/pancakeswapV2-restricted/PancakeswapV2RestrictedStrategyAddBaseTokenOnly.sol [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  mapping(address => bool) public okWorkers;
+
+  /// @dev Require that the caller must be an EOA account to avoid flash loans.
+  modifier onlyWhitelistedWorkers() {
+    require(okWorkers[msg.sender], "PancakeswapV2RestrictedStrategyAddBaseTokenOnly::onlyWhitelistedWorkers:: bad worker");
+    _;
+  } 
+
+  /// @dev Create a new add Token only strategy instance.
+  /// @param _router The Uniswap router smart contract.
```
Comment (spicysquid168) [inline contracts/6/protocol/strategies/pancakeswapV2-restricted/PancakeswapV2RestrictedStrategyAddBaseTokenOnly.sol:33]: Update comment

File: contracts/6/protocol/strategies/pancakeswapV2-restricted/PancakeswapV2RestrictedStrategyAddBaseTokenOnly.sol [hunk 2]
```diff
@@ -0,0 +1,92 @@
+pragma solidity 0.6.6;
```
Comment (spicysquid168) [inline contracts/6/protocol/strategies/pancakeswapV2-restricted/PancakeswapV2RestrictedStrategyAddBaseTokenOnly.sol:1]: Update to 0.6.12

File: contracts/6/protocol/strategies/pancakeswapV2-restricted/PancakeswapV2RestrictedStrategyAddTwosidesOptimal.sol [hunk 3]
```diff
@@ -0,0 +1,140 @@
+pragma solidity 0.6.6;
```
Comment (spicysquid168) [inline contracts/6/protocol/strategies/pancakeswapV2-restricted/PancakeswapV2RestrictedStrategyAddTwosidesOptimal.sol:1]: 0.6.12

File: contracts/6/protocol/strategies/pancakeswapV2-restricted/PancakeswapV2RestrictedStrategyAddTwosidesOptimal.sol [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+  using SafeToken for address;
+  using SafeMath for uint256;
+
+  IPancakeFactory public factory;
+  IPancakeRouter02 public router;
+  IVault public vault;
+
+  mapping(address => bool) public okWorkers;
+
+  /// @dev Require that the caller must be an EOA account to avoid flash loans.
```
Comment (spicysquid168) [inline contracts/6/protocol/strategies/pancakeswapV2-restricted/PancakeswapV2RestrictedStrategyAddTwosidesOptimal.sol:28]: Update comment

File: contracts/6/protocol/strategies/pancakeswapV2-restricted/PancakeswapV2RestrictedStrategyAddTwosidesOptimal.sol [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    baseToken.safeApprove(address(router), 0);
+  }
+
+  function setWorkersOk(address[] calldata workers, bool isOk) external onlyOwner {
+    for (uint256 idx = 0; idx < workers.length; idx++) {
+      okWorkers[workers[idx]] = isOk;
+    }
+  }
+
+  receive() external payable {}
```
Comment (spicysquid168) [inline contracts/6/protocol/strategies/pancakeswapV2-restricted/PancakeswapV2RestrictedStrategyAddTwosidesOptimal.sol:139]: Remove this as we won't received any native in this strat

File: contracts/6/protocol/strategies/pancakeswapV2-restricted/PancakeswapV2RestrictedStrategyLiquidate.sol [hunk 6]
```diff
@@ -0,0 +1,78 @@
+pragma solidity 0.6.6;
```
Comment (spicysquid168) [inline contracts/6/protocol/strategies/pancakeswapV2-restricted/PancakeswapV2RestrictedStrategyLiquidate.sol:1]: 0.6.12

File: contracts/6/protocol/strategies/pancakeswapV2-restricted/PancakeswapV2RestrictedStrategyLiquidate.sol [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+contract PancakeswapV2RestrictedStrategyLiquidate is OwnableUpgradeSafe, ReentrancyGuardUpgradeSafe, IStrategy {
+  using SafeToken for address;
+
+  IPancakeFactory public factory;
+  IPancakeRouter02 public router;
+
+  mapping(address => bool) public okWorkers;
+
+  /// @dev Require that the caller must be an EOA account to avoid flash loans.
```
Comment (spicysquid168) [inline contracts/6/protocol/strategies/pancakeswapV2-restricted/PancakeswapV2RestrictedStrategyLiquidate.sol:23]: Update comment

File: contracts/6/protocol/strategies/pancakeswapV2-restricted/PancakeswapV2RestrictedStrategyPartialCloseLiquidate.sol [hunk 8]
```diff
@@ -0,0 +1,81 @@
+pragma solidity 0.6.6;
```
Comment (spicysquid168) [inline contracts/6/protocol/strategies/pancakeswapV2-restricted/PancakeswapV2RestrictedStrategyPartialCloseLiquidate.sol:1]: Update to 0.6.12

File: contracts/6/protocol/strategies/pancakeswapV2-restricted/PancakeswapV2RestrictedStrategyPartialCloseLiquidate.sol [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+contract PancakeswapV2RestrictedStrategyPartialCloseLiquidate is OwnableUpgradeSafe, ReentrancyGuardUpgradeSafe, IStrategy {
+  using SafeToken for address;
+
+  IPancakeFactory public factory;
+  IPancakeRouter02 public router;
+
+  mapping(address => bool) public okWorkers;
+
+  /// @dev Require that the caller must be an EOA account to avoid flash loans.
```
Comment (spicysquid168) [inline contracts/6/protocol/strategies/pancakeswapV2-restricted/PancakeswapV2RestrictedStrategyPartialCloseLiquidate.sol:23]: Update comment

File: contracts/6/protocol/strategies/pancakeswapV2-restricted/PancakeswapV2RestrictedStrategyWithdrawMinimizeTrading.sol [hunk 10]
```diff
@@ -0,0 +1,103 @@
+pragma solidity 0.6.6;
```
Comment (spicysquid168) [inline contracts/6/protocol/strategies/pancakeswapV2-restricted/PancakeswapV2RestrictedStrategyWithdrawMinimizeTrading.sol:1]: 0.6.12

File: test/PancakeswapV2_Restricted_StrategyAddTwoSidesOptimal.test.ts [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+      await baseToken.mint(mockPancakeswapV2Worker.address, ethers.utils.parseEther('0.1'))
+      await mockPancakeswapV2WorkerAsAlice.work(0, await alice.getAddress(), ethers.utils.parseEther('0.1'), ethers.utils.defaultAbiCoder.encode(
+        ['address', 'bytes'],
+        [addRestrictedStrat.address, ethers.utils.defaultAbiCoder.encode(
+          ['uint256','uint256'],
+          ['0', ethers.utils.parseEther('0')]
+        )],
+      ))
+  
+      expect(await lpV2.balanceOf(mockPancakeswapV2Worker.address)).to.above(stratLPBalance);
```
Comment (spicysquid168) [inline test/PancakeswapV2_Restricted_StrategyAddTwoSidesOptimal.test.ts:270]: add expect to the exact amount with calculation as a comment here

File: test/PancakeswapV2_Restricted_StrategyAddTwoSidesOptimal.test.ts [hunk 12]
```diff
@@ -0,0 +1,345 @@
...
+      await farmingTokenAsAlice.approve(mockedVault.address, ethers.utils.parseEther('1'));
+      await mockPancakeswapV2WorkerAsAlice.work(0, await alice.getAddress(), ethers.utils.parseEther('1'), ethers.utils.defaultAbiCoder.encode(
+        ['address', 'bytes'],
+        [addRestrictedStrat.address, ethers.utils.defaultAbiCoder.encode(
+          ['uint256','uint256'],
+          [ethers.utils.parseEther('1'), ethers.utils.parseEther('0.1')]
+        )],
+      ))
+  
+      expect(await lpV2.balanceOf(mockPancakeswapV2Worker.address)).to.above(stratLPBalance);
```
Comment (spicysquid168) [inline test/PancakeswapV2_Restricted_StrategyAddTwoSidesOptimal.test.ts:306]: add expect to the exact amount with calculation as a comment here

**Review by spicysquid168 — APPROVED** [review]

**Review by almost-llamas — APPROVED** [review]

**Pull request merged**